### DISCUSSION TITLE
Keep mesh centered without pinning nodes?

### DISCUSSION THREAD

**Discussion by jessecarterMOOSE** [idaholab/moose#23063] [Q&A Modules: Solid mechanics] [answered]
I'm trying to recreate some Abaqus results in MOOSE. To more easily measure model distortions, the mesh is kept centered by applying constraints that force the displacements of certain nodes (corner nodes for example) to sum to zero. In Abaqus you can also explicitly set the rotational degrees of freedom to zero so there is no rigid body motion.

I have written a NodalKernel for MOOSE that does the mesh centering, but some rotation can still occur. Is there a way I can prevent rotations without pinning nodes?

**Comment by GiudGiud** [comment, marked as answer]
Hello

Could you follow the same approach and force some measure of the displacement from a rotation (maybe |disp| to avoid compensation between opposite nodes) to sum to zero?

Guillaume

**Reply by jessecarterMOOSE** [reply]
Sounds reasonable but how do you measure the rotation?

**Reply by GiudGiud** [reply]
Oh it s not always around the origin that you forced?

**Reply by jessecarterMOOSE** [reply]
no, the meshes I'm working with are all over the place.

**Reply by GiudGiud** [reply]
is the rotation always around the center of mass though?
is the axis always the same too?

Does this all depend on which nodes you chose to pin by forcing the displacement sum to 0?

**Reply by jessecarterMOOSE** [reply]
I've only inspected it visually, but it does seem the rotation is about the center. Just looking at 2D models for now. Haven't played around with the constraint nodes since those are already built in to the Abaqus models I'm working with.

**Reply by GiudGiud** [reply]
if you know that center you could compute a moment with the displacement for the constraint node and try to constain that.

**Reply by jessecarterMOOSE** [reply]
I do know the point coordinates (which are also a nodeset), so computing the centroid is no problem.

Guess I would need to find the component of the displacement that is perpendicular to the radial vector (center to point), sum those for each point, and constrain that value?

**Reply by GiudGiud** [reply]
I think to compute the moment you would also need to multiply by the radial vector length. 
But both of these are heuristics so either could work?

**Reply by GiudGiud** [reply]
Any luck with this heuristic or the petsc nullspace removal?

**Reply by jessecarterMOOSE** [reply]
I ended up implementing the ad-hoc sum-to-zero constraints I mentioned using NodalKernels for applying forces and ScalarKernels for computing Lagrange multipliers. Multiple constraints were needed to prevent both rotation and translation. Unfortunately my abilities with petsc don't go beyond passing petsc options in the MOOSE input file, so I never tried to tackle the null space removal.

**Comment by hugary1995** [comment]
There's a capability in petsc which automatically computes and removes (near) null spaces in the context of solid mechanics. @reverendbedford recently tried that IIRC.

**Reply by reverendbedford** [reply]
I suppose I should finish that PR...

**Reply by jessecarterMOOSE** [reply]
> There's a capability in petsc which automatically computes and removes (near) null spaces in the context of solid mechanics. @reverendbedford recently tried that IIRC.

@hugary1995 does this capability currently exist in MOOSE?

**Comment by jessecarterMOOSE** [comment]
Let me back up and make sure I'm doing things right. What I'm after is for all $i$ nodes in my nodeset: $\sum_{i} disp_{x,i}=0$, $\sum_{i} disp_{y,i}=0$ and not $disp_{x,i}=0$, $disp_{y,i}=0$.

I'm using `PenaltyDirichletNodalKernel` with `value=0.0`. Is that what I want or do I need to do something different?
